src/main/scala/org/kframework/kore/interfaces/Interface.scala

```diff
@@ -0,0 +1,303 @@
+
+  trait Variable extends Pattern {
+    val name: Name
+    val sort: Sort
+  }
+
+  object Variable {
+    def unapply(arg: Variable): Option[(Name, Sort)] = Some(arg.name, arg.sort)
+  }
+
+  trait Application extends Pattern {
+    val symbol: Symbol
+    val args: Seq[Pattern]
+  }
+
+  object Application {
+    def unapply(arg: Application): Option[(Symbol, Seq[Pattern])] = Some(arg.symbol, arg.args)
+  }
+
+  trait DomainValue extends Pattern {
+    val symbol: Symbol
+    val value: Value
+  }
+
+  object DomainValue {
+    def unapply(arg: DomainValue): Option[(Symbol, Value)] = Some(arg.symbol, arg.value)
+  }
+
+  trait Top extends Pattern
+
+  object Top {
+    def unapply(arg: Top): Boolean = true
+  }
+
+  trait Bottom extends Pattern
+
+  object Bottom {
+    def unapply(arg: Bottom): Boolean = true
+  }
+
+  trait And extends Pattern {
+    val p: Pattern
+    val q: Pattern
+  }
+
+  object And {
+    def unapply(arg: And): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
+  }
+
+  trait Or extends Pattern {
+    val p: Pattern
+    val q: Pattern
+  }
+
+  object Or {
+    def unapply(arg: Or): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
+  }
+
+  trait Not extends Pattern {
+    def p: Pattern
+  }
+
+  object Not {
+    def unapply(arg: Not): Option[Pattern] = Some(arg.p)
+  }
+
+  trait Implies extends Pattern {
+    val p: Pattern
+    val q: Pattern
+  }
+
+  object Implies {
+    def unapply(arg: Implies): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
+  }
+
+  trait Exists extends Pattern {
+    val v: Variable
+    val p: Pattern
+  }
+
+  object Exists {
+    def unapply(arg: Exists): Option[(Variable, Pattern)] = Some(arg.v, arg.p)
+  }
+
+  trait ForAll extends Pattern {
+    val v: Variable
+    val p: Pattern
+  }
+
+  object ForAll {
+    def unapply(arg: ForAll): Option[(Variable, Pattern)] = Some(arg.v, arg.p)
+  }
+
+  trait Next extends Pattern {
+    def p: Pattern
+  }
+
+  object Next {
+    def unapply(arg: Next): Option[Pattern] = Some(arg.p)
+  }
+
+
+  trait Rewrite extends Pattern {
+    val p: Pattern
+    val q: Pattern
+  }
+
+  object Rewrite {
+    def unapply(arg: Rewrite): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
+  }
+
+  trait Equals extends Pattern {
+    val p: Pattern
+    val q: Pattern
+  }
+
+  object Equals {
+    def unapply(arg: Equals): Option[(Pattern, Pattern)] = Some(arg.p, arg.q)
+  }
+
+  trait ModuleName {
+    def str: String
+  }
+
+  object ModuleName {
+    def unapply(arg: ModuleName): Option[String] = Some(arg.str)
+  }
+
+  trait Name {
+    val str: String
+  }
+
+  object Name {
+    def unapply(arg: Name): Option[String] = Some(arg.str)
+  }
+
+  trait Sort {
+    val str: String
+  }
+
+  object Sort {
+    def unapply(arg: Sort): Option[String] = Some(arg.str)
+  }
+
+  trait Symbol {
+    val str: String
+  }
+
+  object Symbol {
+    def unapply(arg: Symbol): Option[String] = Some(arg.str)
+  }
+
+  trait Value {
+    val str: String
```

Comment:
Or even not have the trait to begin with. We started with creating a special wrapper for symbols as it ensured we mess up the order when calling the `DomainValue` constructor, i.e., when writing `DomainValue("foo", "bar")` it's not clear which one is the symbol and which one is the value, and it's very easy to introduce bugs this way (I did). But we could very well live with `DomainValue(Symbol("foo"), "bar")`, or with a `Value` trait that has a String-based builder in `Builders` but relies on the `toString` method here.